network/test/epochtransition_test.go

```diff
@@ -175,38 +280,125 @@ func checkConnectivity(t *testing.T, mw *p2p.Middleware, ids flow.IdentityList)
-	optionalSleep(send)
+// exchangeMessages verifies that allowed engines can successfully exchange messages between them while disallowed
+// engines can't using the ConduitSendWrapperFunc network primitive
+func (suite *MutableIdentityTableSuite) exchangeMessages(
+	allowedIDs flow.IdentityList,
+	allowedEngs []*MeshEngine,
+	disallowedIDs flow.IdentityList,
+	disallowedEngs []*MeshEngine,
+	send ConduitSendWrapperFunc,
+	expectSendErrorForDisallowedIDs bool) {
 
-	count := len(engs)
-	expectedMsgCnt := count - 1
+	// send a message from each of the allowed engine to the other allowed engines
+	for i, allowedEng := range allowedEngs {
 
-	// each node broadcasting a message to all others
-	for i, eng := range engs {
-		event := &message.TestMessage{
-			Text: fmt.Sprintf("hello from node %d", i),
+		fromID := allowedIDs[i].NodeID
+		targetIDs := allowedIDs.Filter(filter.Not(filter.HasNodeID(allowedIDs[i].NodeID)))
+
+		err := suite.sendMessage(fromID, allowedEng, targetIDs, send)
+		require.NoError(suite.T(), err)
+	}
+
+	// send a message from each of the allowed engine to all of the disallowed engines
+	if len(disallowedEngs) > 0 {
+		for i, fromEng := range allowedEngs {
+
+			fromID := allowedIDs[i].NodeID
+			targetIDs := disallowedIDs
+
+			err := suite.sendMessage(fromID, fromEng, targetIDs, send)
+			if expectSendErrorForDisallowedIDs {
+				require.Error(suite.T(), err)
+			}
+		}
+	}
+
+	// send a message from each of the disallowed engine to each of the allowed engines
+	for i, fromEng := range disallowedEngs {
+
+		fromID := disallowedIDs[i].NodeID
+		targetIDs := allowedIDs
+
+		err := suite.sendMessage(fromID, fromEng, targetIDs, send)
+		if expectSendErrorForDisallowedIDs {
+			require.Error(suite.T(), err)
 		}
-		others := ids.Filter(filter.Not(filter.HasNodeID(ids[i].NodeID))).NodeIDs()
-		require.NoError(t, send(event, eng.con, others...))
 	}
 
+	count := len(allowedEngs)
+	expectedMsgCnt := count - 1
 	wg := sync.WaitGroup{}
-	// fires a goroutine for each engine to listen to incoming messages
-	for i := range engs {
+	// fires a goroutine for each of the allowed engine to listen for incoming messages
+	for i := range allowedEngs {
 		wg.Add(expectedMsgCnt)
 		go func(e *MeshEngine) {
 			for x := 0; x < expectedMsgCnt; x++ {
 				<-e.received
 				wg.Done()
 			}
-		}(engs[i])
+		}(allowedEngs[i])
+	}
+
+	// assert that all allowed engines received expectedMsgCnt number of messages
+	unittest.AssertReturnsBefore(suite.T(), wg.Wait, 5*time.Second)
+	// assert that all allowed engines received no other messages
+	for i := range allowedEngs {
+		assert.Empty(suite.T(), allowedEngs[i].received)
+	}
+
+	// assert that the disallowed engines didn't receive any message
+	for i, eng := range disallowedEngs {
+		unittest.RequireNeverReturnBefore(suite.T(), func() {
+			<-eng.received
+		}, time.Millisecond, fmt.Sprintf("%s engine should not have recevied message", disallowedIDs[i]))
```

Comment:
you may use:

```
unittest.RequireCloseBefore(suite.T(), eng.received, fmt.Sprintf("%s engine should not have recevied message", disallowedIDs[i])) 
```

Reply:
here I am trying to test the reverse - eng.received should never be closed